File: app/src/main/java/com/donnfelker/kotlinmix/TodoAdapter.java
package com.donnfelker.kotlinmix;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.donnfelker.kotlinmix.models.Todo;

import io.realm.RealmBasedRecyclerViewAdapter;
import io.realm.RealmResults;
import io.realm.RealmViewHolder;


import butterknife.Bind;
import butterknife.ButterKnife;

public class TodoAdapter extends RealmBasedRecyclerViewAdapter<Todo, TodoAdapter.ViewHolder> {

    private TodoItemClickListener clickListener;

    public TodoAdapter(Context context, RealmResults<Todo> realmResults, boolean automaticUpdate, boolean animateResults, TodoItemClickListener clickListener) {
        super(context, realmResults, automaticUpdate, animateResults);
        this.clickListener = clickListener;
    }

    @Override
    public ViewHolder onCreateRealmViewHolder(ViewGroup viewGroup, int viewType) {
        View v = inflater.inflate(R.layout.todo_item, viewGroup, false);
        return new ViewHolder(v, clickListener);
    }

    @Override
    public void onBindRealmViewHolder(ViewHolder viewHolder, int position) {
        final Todo todo = realmResults.get(position);
        viewHolder.todoTitle.setText(todo.getTitle());
    }

    class ViewHolder extends RealmViewHolder implements View.OnClickListener {

        @Bind(R.id.todo_item_todo_title) public TextView todoTitle;
        private TodoItemClickListener clickListener;

        public ViewHolder(View view, TodoItemClickListener clickListener) {
            super(view);
            this.clickListener = clickListener;
            ButterKnife.bind(this, view);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if(clickListener != null) {
                clickListener.onTodoClick(v, realmResults.get(getAdapterPosition()));
            }
        }
    }

    public interface TodoItemClickListener {
        void onTodoClick(View caller, Todo task);
    }
}

File: app/src/main/kotlin/com/donnfelker/kotlinmix/TodoAdapter.kt
package com.donnfelker.kotlinmix

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.donnfelker.kotlinmix.models.Todo

import io.realm.RealmBasedRecyclerViewAdapter
import io.realm.RealmResults
import io.realm.RealmViewHolder

import butterknife.ButterKnife
import org.jetbrains.anko.find


class TodoAdapter(context: Context, realmResults: RealmResults<Todo>, automaticUpdate: Boolean, animateResults: Boolean, private val clickListener: TodoItemClickListener) : RealmBasedRecyclerViewAdapter<Todo, TodoAdapter.ViewHolder>(context, realmResults, automaticUpdate, animateResults) {
    override fun onCreateRealmViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        val v = inflater.inflate(R.layout.todo_item, viewGroup, false)
        return ViewHolder(v, clickListener)
    }

    override fun onBindRealmViewHolder(viewHolder: ViewHolder, position: Int) {
        val todo = realmResults[position]
        viewHolder.todoTitle!!.text = todo.title
    }

    inner class ViewHolder(view: View, private val clickListener: TodoItemClickListener?) : RealmViewHolder(view), View.OnClickListener {

        val todoTitle: TextView? = view.find<TextView>(R.id.todo_item_todo_title)

        init {
            ButterKnife.bind(this, view)
            view.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            clickListener?.onTodoClick(v, realmResults[adapterPosition])
        }
    }

    interface TodoItemClickListener {
        fun onTodoClick(caller: View, task: Todo)
    }
}